src/ripple/basics/impl/Time.cpp

```diff
@@ -70,6 +70,37 @@ to_string(std::chrono::system_clock::time_point tp)
         << setw(2) << h.count() << ':'
         << setw(2) << m.count() << ':'
         << setw(2) << s.count();
+
+    switch (period)
+    {
+        case Period::none:
+        case Period::seconds:
+            break;
+        case Period::milliseconds:
+        {
+            std::uint64_t totalMs =
+                floor<milliseconds>(tp.time_since_epoch()).count();
+            str << '.' << totalMs - (totalMs / 1000 * 1000);
+        }
+            break;
+        case Period::microseconds:
+        {
+            std::uint64_t totalUs =
+                floor<microseconds>(tp.time_since_epoch()).count();
+            str << '.' << totalUs - (totalUs / 1000000 * 1000000);
+        }
+            break;
+        case Period::nanoseconds:
+        {
+            std::uint64_t totalNs =
+                floor<nanoseconds>(tp.time_since_epoch()).count();
+            str << '.' << totalNs - (totalNs / 1000000000 * 1000000000);
+        }
+            break;
+        default:
+            assert(false);
+    }
+
     return str.str();
 }
```

Comment:
As it stands, this gives the wrong answer.  But it is easy to correct.  What is incorrect is that if there are less than 100ms (for example), the leading `0` in the milliseconds amount is omitted.  I also have an allergy to explicit conversion constants in chrono code.  Both of these can be corrected like this:
```c++
std::string
to_string(std::chrono::system_clock::time_point tp, Period period)
{
    const char* months[] =
        {
            "Jan", "Feb", "Mar", "Apr", "May", "Jun",
            "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"
        };
    using namespace std::chrono;
    auto totalsec = floor<seconds>(tp.time_since_epoch());
    auto s = totalsec;
    auto sd = floor<days>(s);  // number of days
    s -= sd;  // time of day in seconds
    auto h = floor<hours>(s);
    s -= h;
    auto m = floor<minutes>(s);
    s -= m;
    int y;
    unsigned mn, d;
    std::tie(y, mn, d) = civil_from_days(static_cast<int>(sd.count()));
    // Date-time in y/mn/d h:m:s
    std::ostringstream str;
    str.fill('0');
    str.flags(std::ios::dec | std::ios::right);
    using std::setw;
    str << y << '-' << months[mn-1] << '-' << setw(2) << d << ' '
        << setw(2) << h.count() << ':'
        << setw(2) << m.count() << ':'
        << setw(2) << s.count();

    switch (period)
    {
        case Period::none:
        case Period::seconds:
            break;
        case Period::milliseconds:
            str << '.' << setw(3)
                << (floor<milliseconds>(tp.time_since_epoch()) - totalsec).count();
            break;
        case Period::microseconds:
            str << '.' << setw(6)
                << (floor<microseconds>(tp.time_since_epoch()) - totalsec).count();
            break;
        case Period::nanoseconds:
            str << '.' << setw(9)
                << (floor<nanoseconds>(tp.time_since_epoch()) - totalsec).count();
            break;
        default:
            assert(false);
    }

    return str.str();
}
```
There is also an alternative that I'll mention, though it can be seen as overkill.  Using my datetime lib (and likely C++20 at this point), `to_string` can be rewritten like this:
```c++
template <class Duration>
std::string
to_string(date::sys_time<Duration> tp)
{
    return date::format("%Y-%b-%d %T", tp);
}
```
This would require changes at 4 call sites to `floor` the desired precision at the call site.  For example in src/ripple/app/misc/NetworkOPs.cpp
```c++
info [jss::time] = to_string(std::chrono::system_clock::now(), Period::microseconds);
```
would become:
```c++
using namespace std::chrono;
info [jss::time] = to_string(floor<microseconds>(system_clock::now()));
```
If we want to go this route, it requires bringing in date.h from https://github.com/HowardHinnant/date/blob/master/include/date/date.h (and no other sources), and removing our own version of `floor`, `ceil`, `round`.